🌐 Add Ukrainian translation for `docs/uk/docs/index.md` (#10362)
fastapi/fastapi: docs/uk/docs/index.md

<p align="center">
  <a href="https://fastapi.tiangolo.com"><img src="https://fastapi.tiangolo.com/img/logo-margin/logo-teal.png" alt="FastAPI"></a>
</p>
<p align="center">

>     <em>Фреймворк FastAPI, висока продуктивність, простий у вивченні, швидкий для написання коду, готовий до продакшину</em>

Reviewer: ```suggestion
    <em>Фреймворк FastAPI, висока продуктивність, простий у вивченні, швидкий для написання коду, готовий до продакшину</em>
```

Reviewer: this does not loooks good

Author: @akarev0 As I understood these are keywords at the top of the page. Not sentences

</p>
<p align="center">
<a href="https://github.com/tiangolo/fastapi/actions?query=workflow%3ATest+event%3Apush+branch%3Amaster" target="_blank">
    <img src="https://github.com/tiangolo/fastapi/workflows/Test/badge.svg?event=push&branch=master" alt="Test">
</a>
<a href="https://coverage-badge.samuelcolvin.workers.dev/redirect/tiangolo/fastapi" target="_blank">
    <img src="https://coverage-badge.samuelcolvin.workers.dev/tiangolo/fastapi.svg" alt="Coverage">
</a>
<a href="https://pypi.org/project/fastapi" target="_blank">
    <img src="https://img.shields.io/pypi/v/fastapi?color=%2334D058&label=pypi%20package" alt="Package version">
</a>
<a href="https://pypi.org/project/fastapi" target="_blank">
    <img src="https://img.shields.io/pypi/pyversions/fastapi.svg?color=%2334D058" alt="Supported Python versions">
</a>
</p>

---

**Документація**: <a href="https://fastapi.tiangolo.com" target="_blank">https://fastapi.tiangolo.com</a>

**Програмний код**: <a href="https://github.com/tiangolo/fastapi" target="_blank">https://github.com/tiangolo/fastapi</a>

---

> FastAPI це сучасний, швидкий (високопродуктивний), вебфреймворк для створення API за допомогою Python 3.7+ на основі стандартних підказок типу Python.

Reviewer: ```suggestion
FastAPI це сучасний, швидкий (високопродуктивний), вебфреймворк для створення API за допомогою Python 3.7+ на основі стандартних підказок типу Python.
```

Reviewer: думаю що тут `**based on standard Python type hints.**` не конкретно про типи йдеться, бо такий переклад не має сенсу не жаль

Author: @akarev0  я думаю типи тут все ж мають значення, і це відсилка до пайдантика, можливо просто потрібно змінити формулювання речення, щоб воно було більш правильним. Є якісь пропозиції?

Ключові особливості:

* **Швидкий**: Дуже висока продуктивність, на рівні з **NodeJS** та **Go** (завдяки Starlette та Pydantic). [Один із найшвидших фреймворків](#performance).

> * **Швидке написання коду**: Збільшить швидкість розробки функціоналу приблизно на 200%-300%. *

Reviewer: ```suggestion
* **Швидке написання коду**: Пришвидшує розробку функціоналу приблизно на 200%-300%. *
```

* **Менше помилок**: Зменшить кількість помилок спричинених людиною (розробником) на 40%. *
* **Інтуїтивний**: Чудова підтримка редакторами коду. <abbr title="Також відоме як auto-complete, autocompletion, IntelliSense.">Доповнення</abbr> всюди. Зменште час на налагодження.
* **Простий**: Спроектований, для легкого використання та навчання. Знадобиться менше часу на читання документації.
* **Короткий**: Зведе до мінімуму дублювання коду. Кожен оголошений параметр може виконувати кілька функцій.

> * **Надійний**: Ви матимете стабільний код готовий для продакшину з автоматичною інтерактивною документацією.

Reviewer 2: ```suggestion
* **Надійний**: Ви матимете стабільний код готовий до продакшину з автоматичною інтерактивною документацією.
```

<div style="text-align: right; margin-right: 10%;">Timothy Crosley - <strong><a href="https://www.hug.rest/" target="_blank">Hug</a> creator</strong> <a href="https://news.ycombinator.com/item?id=19455465" target="_blank"><small>(ref)</small></a></div>

---

"_If you're looking to learn one **modern framework** for building REST APIs, check out **FastAPI** [...] It's fast, easy to use and easy to learn [...]_"

"_We've switched over to **FastAPI** for our **APIs** [...] I think you'll like it [...]_"

<div style="text-align: right; margin-right: 10%;">Ines Montani - Matthew Honnibal - <strong><a href="https://explosion.ai" target="_blank">Explosion AI</a> founders - <a href="https://spacy.io" target="_blank">spaCy</a> creators</strong> <a href="https://twitter.com/_inesmontani/status/1144173225322143744" target="_blank"><small>(ref)</small></a> - <a href="https://twitter.com/honnibal/status/1144031421859655680" target="_blank"><small>(ref)</small></a></div>

---

## **Typer**, FastAPI CLI

<a href="https://typer.tiangolo.com" target="_blank"><img src="https://typer.tiangolo.com/img/logo-margin/logo-margin-vector.svg" style="width: 20%;"></a>

> Якщо ви створюєте <abbr title="Command Line Interface">CLI</abbr> застосунок для використання в терміналі, замість веб-API зверніть увагу на <a href="https://typer.tiangolo.com/" class="external-link" target="_blank">**Typer**</a>.

Reviewer: ```suggestion
Створюючи <abbr title="Command Line Interface">CLI</abbr> застосунок для використання в терміналі, замість веб-API зверніть увагу на <a href="https://typer.tiangolo.com/" class="external-link" target="_blank">**Typer**</a>.
```

**Typer** є молодшим братом FastAPI. І це **FastAPI для CLI**. ⌨️ 🚀

## Вимоги

> Python 3.7+

Reviewer 3: https://github.com/tiangolo/fastapi/releases/tag/0.104.0 dropped the support for 3.7, now it's 3.8+.

Author: @OlegZv Thanks, updated.  PR was opened before this changes)

$ pip install fastapi

---> 100%
```

</div>

Вам також знадобиться сервер ASGI для продакшину, наприклад <a href="https://www.uvicorn.org" class="external-link" target="_blank">Uvicorn</a> або <a href="https://gitlab.com/pgjones/hypercorn" class="external-link" target="_blank">Hypercorn</a>.

<div class="termy">

```console
$ pip install uvicorn[standard]

---> 100%
```

</div>

## Приклад

### Створіть

* Створіть файл `main.py` з:

```Python
from typing import Union

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```

<details markdown="1">
<summary>Або використайте <code>async def</code>...</summary>

Якщо ваш код використовує `async` / `await`, скористайтеся `async def`:

```Python hl_lines="9  14"
from typing import Union

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
async def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```

**Примітка**:

> Якщо ви не розумієте, перегляньте розділ _"In a hurry?"_ про <a href="https://fastapi.tiangolo.com/async/#in-a-hurry" target="_blank">`async` та `await` в документації</a>.

Reviewer: ```suggestion
Якщо ви не розумієте, перегляньте розділ _"In a hurry?"_ про <a href="https://fastapi.tiangolo.com/async/#in-a-hurry" target="_blank">`async` та `await` у документації</a>.
```

> Якщо ви не розумієте, перегляньте розділ _"In a hurry?"_ про <a href="https://fastapi.tiangolo.com/async/#in-a-hurry" target="_blank">`async` та `await` в документації</a>.

Reviewer: ```suggestion
Стикнувшись з проблемами, не зайвим буде ознайомитися з розділом _"In a hurry?"_ про <a href="https://fastapi.tiangolo.com/async/#in-a-hurry" target="_blank">`async` та `await` в документації</a>.
```

Author: @akarev0 "що стосується асінк та авейт в документації" , не звучить) потрібно або залишати `про`, або змінювати і частину `в документації`

Reviewer: "в документації" не можу бути бо дві приголосні в ряд погано читаються. Там можна і так залишити але оце я б точно змінив
```Якщо ви не розумієте, перегляньте розділ```

Author: @akarev0 я не про прийменник у/в, тут питань не має, я про нове речення в цілому

</details>

### Запустіть

Запустіть server з:

<div class="termy">

```console
$ uvicorn main:app --reload

INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [28720]
INFO:     Started server process [28722]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

</div>

<details markdown="1">
<summary>Про команди <code>uvicorn main:app --reload</code>...</summary>

Команда `uvicorn main:app` посилається на:

* `main`: файл `main.py` ("Модуль" Python).
* `app`: об’єкт створений усередині `main.py` рядком `app = FastAPI()`.
* `--reload`: перезапускає сервер після зміни коду. Використовуйте виключно для розробки.

</details>

### Перевірте

Відкрийте браузер та введіть адресу <a href="http://127.0.0.1:8000/items/5?q=somequery" class="external-link" target="_blank">http://127.0.0.1:8000/items/5?q=somequery</a>.

Ви побачите у відповідь подібний JSON:

```JSON
{"item_id": 5, "q": "somequery"}
```

> Ви вже створили API який:

Reviewer 4: ```suggestion
Ви вже створили API, який:
```

* Отримує HTTP запити за _шляхами_ `/` та `/items/{item_id}`.
* Обидва _шляхи_ приймають `GET` <em>операції</em> (також відомі як HTTP _методи_).

> * _Шлях_ `/items/{item_id}` має _параметер шляху_ `item_id` який має бути `int`.

Reviewer: ```suggestion
* _Шлях_ `/items/{item_id}` має _параметр шляху_ `item_id` який має бути `int`.
```

> * _Шлях_ `/items/{item_id}` має опціональний `str` _параметр запиту_ `q`.

Reviewer: ```suggestion
* _Шлях_ `/items/{item_id}` містить необов'язковий `str` _параметр запиту_ `q`.
```

### Інтерактивні документації API

Перейдемо сюди <a href="http://127.0.0.1:8000/docs" class="external-link" target="_blank">http://127.0.0.1:8000/docs</a>.

Ви побачите автоматичну інтерактивну API документацію (створену завдяки <a href="https://github.com/swagger-api/swagger-ui" class="external-link" target="_blank">Swagger UI</a>):

![Swagger UI](https://fastapi.tiangolo.com/img/index/index-01-swagger-ui-simple.png)

### Альтернативні документації API

> Тепер, перейдемо сюди <a href="http://127.0.0.1:8000/redoc" class="external-link" target="_blank">http://127.0.0.1:8000/redoc</a>.

Reviewer 4: ```suggestion
Тепер перейдемо сюди <a href="http://127.0.0.1:8000/redoc" class="external-link" target="_blank">http://127.0.0.1:8000/redoc</a>.
```

Ви побачите альтернативну автоматичну документацію (створену завдяки  <a href="https://github.com/Rebilly/ReDoc" class="external-link" target="_blank">ReDoc</a>):

![ReDoc](https://fastapi.tiangolo.com/img/index/index-02-redoc-simple.png)

## Приклад оновлення

> Тепер модифікуйте файл `main.py` щоб отримати вміст запиту `PUT`.

Reviewer 4: ```suggestion
Тепер модифікуйте файл `main.py`, щоб отримати вміст запиту `PUT`.
```

Оголошуйте вміст запиту за допомогою стандартних типів Python завдяки Pydantic.

```Python hl_lines="4  9-12  25-27"
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    name: str
    price: float
    is_offer: Union[bool, None] = None

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}

@app.put("/items/{item_id}")
def update_item(item_id: int, item: Item):
    return {"item_name": item.name, "item_id": item_id}
```

Сервер повинен автоматично перезавантажуватися (тому що Ви додали `--reload` до `uvicorn` команди вище).

### Оновлення інтерактивної API документації

Тепер перейдемо сюди <a href="http://127.0.0.1:8000/docs" class="external-link" target="_blank">http://127.0.0.1:8000/docs</a>.

> * Інтерактивна документація API буде автоматично оновлена, включаючи нове тіло:

Reviewer: ```suggestion
* Інтерактивна документація API буде автоматично оновлена, включаючи новий вміст:
```

![Swagger UI](https://fastapi.tiangolo.com/img/index/index-03-swagger-02.png)

* Натисніть кнопку "Try it out", це дозволить вам заповнити параметри та безпосередньо взаємодіяти з API:

![Swagger UI interaction](https://fastapi.tiangolo.com/img/index/index-04-swagger-03.png)

* Потім натисніть кнопку "Execute", інтерфейс користувача зв'яжеться з вашим API, надішле параметри, у відповідь отримає результати та покаже їх на екрані:

![Swagger UI interaction](https://fastapi.tiangolo.com/img/index/index-05-swagger-04.png)

### Оновлення альтернативної API документації

Зараз перейдемо <a href="http://127.0.0.1:8000/redoc" class="external-link" target="_blank">http://127.0.0.1:8000/redoc</a>.

* Альтернативна документація також показуватиме новий параметр і тіло запиту:

![ReDoc](https://fastapi.tiangolo.com/img/index/index-06-redoc-02.png)

### Підсумки

> Таким чином, Ви  **один раз** оголошуєте типи параметрів, тіла, тощо,  як параметри функції

Reviewer 2: ```suggestion
Таким чином, Ви  **один раз** оголошуєте типи параметрів, тіла тощо,  як параметри функції.
```

Ви робите це за допомогою стандартних сучасних типів Python.

Вам не потрібно вивчати новий синтаксис, методи чи класи конкретної бібліотеки тощо.

Використовуючи стандартний **Python 3.7+**.

Наприклад, для `int`:

```Python
item_id: int
```

або для більш складної моделі `Item`:

```Python
item: Item
```

...і з цим єдиним оголошенням Ви отримуєте:

* Підтримку редактора, включаючи:
    * Варіанти заповнення.

>     * Перевірка типів.

Reviewer 4: ```suggestion
    * Перевірку типів.
```

* Перевірку даних:

>     * Автоматичне видалення помилок, коли дані неправильні.

Reviewer: ```suggestion
    * Автоматичні та зрозумілі помилки, у разі некоректних даних.
```

    * Перевірка навіть для JSON з високим рівнем вкладеності.
* <abbr title="також відомий як: serialization, parsing, marshalling">Перетворення</abbr> вхідних даних: з мережі до даних і типів Python. Читання з:
    * JSON.
    * Параметрів шляху.
    * Параметрів запиту.
    * Cookies.
    * Headers.
    * Forms.
    * Файлів.
* <abbr title="також відомий як: serialization, parsing, marshalling">Перетворення</abbr> вихідних даних: з типів і даних Python до мережевих даних (як JSON):
    * Конвертація Python типів (`str`, `int`, `float`, `bool`, `list`, тощо).
    * `datetime` об'єкти.
    * `UUID` об'єкти.
    * Моделі бази даних.
    * ...та багато іншого.

> * Автоматична інтерактивна документація API, включаючи 2 альтернативні інтерфейси користувача:

Reviewer 4: ```suggestion
* Автоматичну інтерактивну документацію API, включаючи 2 альтернативні інтерфейси користувача:
```

    * Swagger UI.
    * ReDoc.

---

Повертаючись до попереднього прикладу коду, **FastAPI**:

* Підтвердить наявність `item_id` у шляху для запитів `GET` та `PUT`.
* Підтвердить, що `item_id` має тип `int` для запитів `GET` and `PUT`.
    * Якщо це не так, клієнт побачить корисну, зрозумілу помилку.
* Перевірить, чи є необов'язковий параметр запиту з назвою `q` (а саме `http://127.0.0.1:8000/items/foo?q=somequery`) для запитів `GET`.
    * Оскільки параметр `q` оголошено як `= None`, він необов'язковий.

>     * За відсутності `None` він був би обов'язковим (як і тіло у випадку з `PUT`).
> * Для запитів `PUT` із `/items/{item_id}`, читає тіло як JSON:

Reviewer: ```suggestion
    * За відсутності `None` він був би обов'язковим (як і вміст у випадку з `PUT`).
* Для запитів `PUT` із `/items/{item_id}`, читає вміст як JSON:
```

>     * Перевірить, чи є обов'язковий атрибут `name` який має бути `str`.
>     * Перевірить, чи є обов'язковий атрибут `price` який має бути `float`.
>     * Перевірить, чи є необов'язковий атрибут `is_offer`, який має бути `bool`, якщо він є.

Reviewer: ```suggestion
    * Перевірить, чи є обов'язковий атрибут `name` тип `str`.
    * Перевірить, чи є обов'язковий атрибут `price` тип `float`.
    * Перевірить, чи є необов'язковий атрибут `is_offer`, тип `bool`, та його наявність.
```

Author: @akarev0 тоді краще `типу` str і тд

Author: чи навіть `типом`. 
Або взагалі змінити структуру речення на щотсь типу: "перевірить, чи задовольняє обовʼязковий атрибут 'name' тип 'str'

    * Усе це також працюватиме для глибоко вкладених об'єктів JSON.
* Автоматично конвертує із та в JSON.
* Документує все за допомогою OpenAPI, який може бути використано в:
    * Інтерактивних системах документації.
    * Системах автоматичної генерації клієнтського коду для багатьох мов.
* Надає безпосередньо 2 вебінтерфейси інтерактивної документації.

---

> Ми лише трошки доторкнулися, але ви вже маєте уявлення про те, як все працює.

Reviewer: ```suggestion
Ми лише трошки доторкнулися, але Ви вже маєте уявлення про те, як все працює.
```

> Ми лише трішки доторкнулися, але Ви вже маєте уявлення про те, як все працює.

Reviewer 2: ```suggestion
Ми лише трішки доторкнулися до коду, але Ви вже маєте уявлення про те, як все працює.
```

Спробуйте змінити рядок:

```Python
    return {"item_name": item.name, "item_id": item_id}
```

...із:

```Python
        ... "item_name": item.name ...
```

...на:

```Python
        ... "item_price": item.price ...
```

...і побачите, як ваш редактор автоматично заповнюватиме атрибути та знатиме їхні типи:

![editor support](https://fastapi.tiangolo.com/img/vscode-completion.png)

Для більш повного ознайомлення з додатковими функціями, перегляньте <a href="https://fastapi.tiangolo.com/tutorial/">Туторіал - Посібник Користувача</a>.

**Spoiler alert**: туторіал - посібник користувача містить:

* Оголошення **параметрів** з інших місць як: **headers**, **cookies**, **form fields** та **files**.
* Як встановити **перевірку обмежень** як `maximum_length` або `regex`.

> * Дуже потужна і проста у використанні система **<abbr title="також відома як: components, resources, providers, services, injectables">Ін'єкція Залежностей</abbr>**.

Reviewer 4: ```suggestion
* Дуже потужна і проста у використанні система **<abbr title="також відома як: components, resources, providers, services, injectables">Впровадження Залежностей</abbr>**.
```

Author: @romashevchenko In this place, it is very important to use `Інʼєкція`. There is no room for artistic interpretation, as this distorts the one of basic principles of programming and creates some misconceptions.

By the way, all your other comments I fixed. Could you pls approve it?

* Багато додаткових функцій (завдяки Starlette) як-от:
    * **WebSockets**
    * надзвичайно прості тести на основі HTTPX та `pytest`
    * **CORS**
    * **Cookie Sessions**
    * ...та більше.

## Продуктивність

Незалежні тести TechEmpower показують що застосунки **FastAPI**, які працюють під керуванням Uvicorn <a href="https://www.techempower.com/benchmarks/#section=test&runid=7464e520-0dc2-473d-bd34-dbdfd7e85911&hw=ph&test=query&l=zijzen-7" class="external-link" target="_blank">є одними з найшвидших серед доступних фреймворків в Python</a>, поступаючись лише Starlette та Uvicorn (які внутрішньо використовуються в FastAPI). (*)

Щоб дізнатися більше про це, перегляньте розділ <a href="https://fastapi.tiangolo.com/benchmarks/" class="internal-link" target="_blank">Benchmarks</a>.

## Необов'язкові залежності

Pydantic використовує:

* <a href="https://github.com/JoshData/python-email-validator" target="_blank"><code>email_validator</code></a> - для валідації електронної пошти.
* <a href="https://docs.pydantic.dev/latest/usage/pydantic_settings/" target="_blank"><code>pydantic-settings</code></a> - для управління налаштуваннями.

> * <a href="https://docs.pydantic.dev/latest/usage/types/extra_types/extra_types/" target="_blank"><code>pydantic-extra-types</code></a> - для додаткових типів що можуть бути використані з Pydantic.

Reviewer 4: ```suggestion
* <a href="https://docs.pydantic.dev/latest/usage/types/extra_types/extra_types/" target="_blank"><code>pydantic-extra-types</code></a> - для додаткових типів, що можуть бути використані з Pydantic.
```
